src/core/bytecode.c

```diff
@@ -641,7 +641,17 @@ void MVM_bytecode_finish_frame(MVMThreadContext *tc, MVMCompUnit *cu,
             entry->value = j;
 
             sf->body.lexical_types[j] = read_int16(pos, 6 * j);
-            MVM_HASH_BIND(tc, sf->body.lexical_names, name, entry)
+
+            /* If we do not have "many" lexicals, don't bother making a lookup
+             * hash. Instead, MVM_get_lexical_by_name will simply do a linear
+             * search of the list. This should be faster for short lists.
+             * The choice of 5 entries is guesswork, and ought to be refined by
+             * benchmarking. This location is the only place where we need this
+             * "magic" constant, hence I don't see the ned to #define it
```

Comment:
"ned"

Reply:
@lizmat Not sure if I'm answering the wrong question here...

There is a list and (currently) a lookup hash maintained for each and every static frame. The threshold for "when not to bother with a hash" is going to be per-frame. It's easiest to have it as a constant, and the same constant for all frames.

You're right that it should be "becomes slower than a hash lookup" - I have no good idea when that is, or even how to measure it well. 5 was a guess. The memory overhead of the hash structure is a lot more than 5 structs, and there's a bunch more CPU work before you even get to compare the wanted string key with the entry in the hash. Plus doing the hash lookup generates (at least) 1 more pointer indirection, 2 if your hash isn't perfectly spread out and you have two entries in the same linked list.